mock_drivers/rosbag_mock_drivers/include/rosbag_mock_drivers/MockDriver.h

```diff
@@ -0,0 +1,183 @@
+   * \param queue_size The size of the queue to use both for the publisher and subscriber
+   *
+   * TODO: This function can be combined with addPassthroughPubNoHeader using C++ 17's constexpr if statements.
+   *
+   */
+  template <typename T>
+  void addPassthroughPub(const std::string& sub_topic, const std::string& pub_topic, bool latch, size_t queue_size);
+
+  /**
+   * \brief Function adds both a publisher and subscriber of the specified type.
+   *        This means a passthrough subscription has been created.
+   *        This version of the method should be used for message types which does NOT contain a Header at the field
+   * "header"
+   *
+   * \tparam T ROS message type which the provided topic name applies to.
+   *         Should be the base type only such as sensor_msgs/NavSatFix
+   *
+   * \param sub_topic The topic name to subscribe to.
+   * \param pub_topic The topic name to publish to.
+   * \param latch Flag idicating if the output publisher should be latched.
+   * \param queue_size The size of the queue to use both for the publisher and subscriber
+   *
+   * TODO: This function can be combined with addPassthroughPub using C++ 17's constexpr if statements.
+   *
+   */
+  template <typename T>
+  void addPassthroughPubNoHeader(const std::string& sub_topic, const std::string& pub_topic, bool latch,
+                                 size_t queue_size);
+
+public:
+  virtual ~MockDriver() = 0;
+
+  /**
+   *  \brief A function to initialize the publishers and subsricers and start the node.
+   *         In child classes, this function will trigger a call to onRun().
+   *
+   *  \return An integer error code. A non-zero value will indicate an initialization failure and result in node
+   *          shutdown.
+   */
+  int run();
+
+  /**
+   * \brief Pure Virtual method which child classes must override that returns the list of all driver types that class
+   * implements
+   *
+   * \return A list of DriverType that defines the interfaces the implementing class provides
+   */
+  virtual std::vector<DriverType> getDriverTypes() = 0;
+
+  /**
+   * \brief Pure Virtual method. Returns an integer value which corresponds to the cav_msgs/DriverStatus enum felids
+   * representing the status of the driver.
+   * 
+   * \return Integer value which must match one of the enum fields in cav_msgs/DriverStatus
+   */
+  virtual uint8_t getDriverStatus() = 0;
+
+  /**
+   * \brief Pure virtual method that returns the desired operational rate of a child class.
+   * 
+   * \return Spin rate in Hz
+   */ 
+  virtual unsigned int getRate() = 0;
+
+  /**
+   * \brief Callback which will be triggered at the rate specified by getRate(). This callback will also trigger the onSpin() method. 
+   * 
+   * \return False if this node should shutdown. True otherwise.
+   */ 
+  bool spinCallback();
+};
+
+}  // namespace mock_drivers
+
+#include "impl/MockDriver.tpp" // Include for templated function implementations
```

Comment:
Can we put this at the top with other includes?

Reply:
Actually no. Since the .tpp file contains the implementation of the functions in this class it has to be placed after the function declarations. This is a pattern for implementation hiding of template functions which must be defined in headers instead of .cpp files.